File: problems/chapter16/miiitomi/004.cpp
// https://onlinejudge.u-aizu.ac.jp/problems/1163

#include <iostream>
#include <vector>
using namespace std;

struct Edge {
    int rev, to, cap;
    Edge(int r, int t, int c) {
        rev = r;
        to = t;
        cap = c;
    }
};

int gcd(int m, int n) {
    if (m < n) swap(m, n);
    if (n == 0) return m;
    return gcd(n, m % n);
}

bool rec(vector<vector<Edge>> &G, vector<bool> &seen, int v) {
    if (seen[v]) return false;
    seen[v] = true;
    if (v == (int)G.size() - 1) return true;

    for (auto &e : G[v]) {
        if (e.cap == 0) continue;

        if (rec(G, seen, e.to)) {
            G[e.to][e.rev].cap++;
            e.cap--;
            return true;
        }
    }
    return false;
}

int FordFulkerson(vector<vector<Edge>> &G) {
    int ans = 0;
    while (true) {
        vector<bool> seen((int)G.size(), false);
        if (rec(G, seen, 0)) ans++;
        else break;
    }
    return ans;
}

int main() {
    vector<int> ans;
    while (true) {
        int m, n;
        cin >> m >> n;
        if (m == 0 && n == 0) break;

        vector<int> b(m), r(n);
        for (int i = 0; i < m; i++) cin >> b[i];
        for (int i = 0; i < n; i++) cin >> r[i];

        vector<vector<Edge>> G(m+n+2);
        for (int i = 1; i <= m; i++) {
            G[0].push_back(Edge(0, i, 1));
            G[i].push_back(Edge(i, 0, 0));
        }
        for (int i = m + 1; i <= m + n; i++) {
            G[i].push_back(Edge(i-m-1, m+n+1, 1));
            G[m+n+1].push_back(Edge(0, i, 0));
        }
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (gcd(b[i], r[j]) == 1) continue;
                int x = i + 1;
                int y = j + m + 1;
                G[x].push_back(Edge((int)G[y].size(), y, 1));
                G[y].push_back(Edge((int)G[x].size()-1, x, 0));
            }
        }

        ans.push_back(FordFulkerson(G));
    }

    for (int x : ans) cout << x << endl;
}

File: problems/chapter16/mshibatatt/004.py
# https://judge.u-aizu.ac.jp/onlinejudge/description.jsp?id=1163&lang=jp
from math import gcd
import sys
sys.setrecursionlimit(10000)

def floyd_warshall(G):
    N = len(G)
    dp = [[float('Inf')] * N for _ in range(N)]
    # init condition
    for i in range(N):
        dp[i][i] = 0
    for i, e in enumerate(G):
        for nxt in e:
            dp[i][nxt[0]] = nxt[1]
    
    for k in range(N):
        for i in range(N):
            for j in range(N):
                dp[i][j] = min(dp[i][j], dp[i][k] + dp[k][j])
    
    return dp

class FordFulkerson:
    def __init__(self, N):
        self.N = N
        self.G = [[] for i in range(N)]
 
    def add_edge(self, fr, to, cap):
        forward = [to, cap, None]
        forward[2] = backward = [fr, 0, forward]
        self.G[fr].append(forward)
        self.G[to].append(backward)
 
    def add_multi_edge(self, v1, v2, cap1, cap2):
        edge1 = [v2, cap1, None]
        edge1[2] = edge2 = [v1, cap2, edge1]
        self.G[v1].append(edge1)
        self.G[v2].append(edge2)
 
    def dfs(self, v, t, f):
        if v == t:
            return f
        seen = self.seen
        seen[v] = True
        for e in self.G[v]:
            w, cap, rev = e
            if cap and not seen[w]:
                d = self.dfs(w, t, min(f, cap))
                if d:
                    e[1] -= d
                    rev[1] += d
                    return d
        return 0
 
    def flow(self, s, t):
        flow = 0
        f = INF = float('inf')
        N = self.N
        while f:
            self.seen = [False]*N
            f = self.dfs(s, t, INF)
            flow += f
        return flow

def main():
    m, n = map(int, input().split())
    if m == n == 0:
        return True
    blue = list(map(int, input().split()))
    red = list(map(int, input().split()))
    
    s = 0
    t = m + n + 1
    ff = FordFulkerson(t+1)
    
    # index of blue cards are i
    for i in range(1, m+1):
        ff.add_edge(0, i, 1)
    # index of red cards are j
    for j in range(m+1, t):
        ff.add_edge(j, t, 1)
            
    for i in range(1, m+1):
        for j in range(m+1, t):
            b = blue[i-1]
            r = red[j-m-1]
            if gcd(b, r) > 1:
                ff.add_edge(i, j, 1)

    output = ff.flow(s, t)

    print(output)
    return False

if __name__ == '__main__':
    while True:
        if main():
            break